### ISSUE TITLE
Question: Recommendations on how to handle dark mode

### ISSUE THREAD

**Issue by orta** [octref/shiki#33]
👋 I'm thinking about how to handle dark and light themes, https://github.com/microsoft/TypeScript-Website/issues/342 eventually I'm going to need to figure out how to make a dark mode which uses css classes because these poor dark mode folks have the same code color schema:

<img width="1458" alt="Screen Shot 2020-04-09 at 9 39 25 PM" src="https://user-images.githubusercontent.com/49038/78954618-b2bd4980-7aaa-11ea-8ff2-5501b5128e4f.png">

<img width="1458" alt="Screen Shot 2020-04-09 at 9 39 17 PM" src="https://user-images.githubusercontent.com/49038/78954646-c36dbf80-7aaa-11ea-9e30-e501a6eee029.png">

So, I'm wondering what you think might be the best way to do it (and whether it's something I should do upstream back here) 

I have two ideas:

 - Have a single theme, convert the colors hex to css classes
 - Try match two separate themes together

**Comment by giovanicascaes** [comment]
Does your first idea mean using a single style that looks good both in light and dark mode?

Also, I'd like to help implementing this, is it something lower in complexity?

**Comment by orta** [comment]
I think of it as being: 

- Shiki returns "#ffeeaa" for a token
- My app makes a sha of it: "#ffeeaa" -> "asd12" 
- My app sets the class of the token to be "asd12"
- We keep track of all colors which were assigned and spit out a single css file for all those colors
-  That CSS file can use `prefers-color-scheme: dark`

**Comment by giovanicascaes** [comment]
It is not an elegant solution but code conteiners could be dark even in light theme. Check Dan Abramov's overreacted.io (https://overreacted.io/goodbye-clean-code/, for example). It looks good for me to read, both in light and dark theme. That seems the easiest possible solution. At least could bring some fresh to the eyes at night until the ideal solution could come up.

**Comment by orta** [comment]
Yep - that's a reasonable workaround to not have this problem. That's totally doable for people, but I'm not planning on making that design compromise

**Comment by giovanicascaes** [comment]
Yes, I understand. I'm not proposing defining as design choice, but it is a better solution than the current one. I believe devs feel more comfortable reading dark code during the day than light code at night.

**Comment by Microflash** [comment]
I don't think it is practical enough to support CSS properties-backed dark and light theme, given the size of grammar and sheer number of color tokens. What I'd suggest is to customize some color tokens (like background-color and foreground-color) through CSS properties and toggle them through a feature query. You'll have to find a theme which has got tokens that work nice enough in light and dark mode (which is annoyingly hard). I was able to do that with Prism (an example [here](https://mflash.dev/blog/2020/04/10/securing-spring-boot-apis-with-jwt-authentication/)), given its very limited set of tokens but gave up on Shiki.

**Comment by orta** [comment]
@pveyes solved this by switching the output to use css variables:
https://fatihkalifa.com/typescript-twoslash

![Screen Shot 2020-05-21 at 3 59 52 PM](https://user-images.githubusercontent.com/49038/82600587-1fcd0000-9b7c-11ea-8954-b2cffdf56e27.png)

**Comment by pveyes** [comment]
Hi, upon closer inspection I found out that we can get `token.explanation` when using `highlighter.codeToThemedTokens`. I think we can implemented dark/light mode support purely in user space without modification in shiki. CMIIW

I'm not really familiar with how vscode textmate parses grammar and the resulting type so here's my assumption. We can find what scope matches inside `token.explanation[0].scopes`

```ts
type Scope = {
  scopeName: string;
  themeMatches: Array<ThemeMatch>
}

const token = highlighter.codeToThemedTokens(code, lang)

const matchingScope: Scope = token.explanation[0].scopes.find(scope => {
  return scope.themeMatches.length > 0
});
```

If empty we can use `--fallback-color`, or other convention

```ts
if (!matchingScope) {
  return "--fallback-color"
}
```

Even though technically we can generate CSS variable using `scope.scopeName`, I think it's better to use `themeMatch` to reduce the number of generated CSS variable.

I found that 0 index is the one that's being used if there's more than 1 `themeMatches`

```ts
type Theme = {
  name: string;
  scope: Array<string>;
}

const matchingTheme: Theme = matchingScope.themeMatches[0]

// we can then generate CSS variable from either name or scope
// remove whitespace, convert invalid character, etc
const cssVar = convertToCSSVariable(matchingTheme.name)

// or use scope
// join all scope, compute hash
const cssVar = convertToCSSVariable(matchingTheme.scopes);

return cssVar
```

Then we can generate HTML tag and use [CSS variable fallback](https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_custom_properties#Custom_property_fallback_values). 

```ts
const cssVar = generateCSSVariable(token);
const html += `<span style="color: var(${cssVar}, ${token.color})">${token.content}</span>`
```

**Comment by Gerrit0** [comment]
So, dark mode alone wasn't quite sufficient for me. I wanted the color switching with themes that @pveyes built, but without being locked into a specific theme / set of themes. I'd love to have readable class names, but after messing around with scopes as suggested above, was unable to come up with a good solution and have settled for just generating `hl-1`, `hl-2`...

Leaving this here as it might be useful to someone - https://gist.github.com/Gerrit0/275a4b8ffee4fa133fd075f5edeb3cda. TypeDoc will use a similar approach in the rebuilt themes.

![image](https://user-images.githubusercontent.com/19329837/86496502-f8508200-bd3a-11ea-8211-b28f5dfa75f7.png)

![image](https://user-images.githubusercontent.com/19329837/86496510-03a3ad80-bd3b-11ea-8ab7-9ca26b07de39.png)

**Comment by octref** [comment]
https://github.com/anotherglitchinthematrix/monochrome could be a good test.

![image](https://user-images.githubusercontent.com/4033249/91000503-e634e680-e5fb-11ea-8dad-9c653f12cd97.png)

I could generate 10 different variations and make a slider demo.

**Comment by orta** [comment]
We are currently inverting and colour hue shifting on the TypeScript website, and I think that's probably enough for us - https://github.com/microsoft/TypeScript-Website/pull/1536

**Comment by hipstersmoothie** [comment]
@orta's approach is an okay solution but I don't think it works for every theme. Works well on the TS website though. My ideal API would be the following:

```js
const highlighter = await shiki.getHighlighter({
  theme: 'github-light',
  darkTheme: 'github-dark',
  langs: [...BUNDLED_LANGUAGES, ...langs]
})
```

As a user I want to be able to pick the theme that best fits my website for light/dark mode.

**Comment by hipstersmoothie** [comment]
I think I've found a pretty good solution. I'm editing `rehype-shiki` to support a `darkTheme` option.
That option will make the rehype plugin generate 2 separate `code` blocks, one for light mode and one for dark mode with a class that reflects that (ex: `syntax-dark`).

Then just by adding the following CSS you can 

```css
.syntax-dark {
  display: none;
}

@media (prefers-color-scheme: dark) {
  .syntax-light {
    display: none;
  }

  .syntax-dark {
    display: block;
  }
}
```

**Comment by orta** [comment]
Hah, nice idea

**Comment by octref** [comment]
I built a playground with dark mode. You can give it a try at https://shiki-play.matsu.io. Source code is at https://github.com/shikijs/shiki-playground

Here are my thoughts on different approaches:

- Outputting two HTML blocks is actually the simplest way as @hipstersmoothie suggested.
- Most themes have a corresponding dark/light theme. I think it's easier for 99% of users to just use the provided counterpart theme, instead of messing with CSS variables and coming up with other colors.

### Outputting semantic HTML (meaningful class names) with CSS

Given a source, a grammar and a theme, you can get the matching scope that makes a token a specific color. For example, here the token is `#9ecbff` because it's `string`.

![image](https://user-images.githubusercontent.com/4033249/113004758-6cbf3100-91a6-11eb-924a-9e26b8de13f2.png)

So we need a renderer that does this:
- For each token with a matching scope `foo.bar` that determines its color
- Give it a class `foo-bar` (convert all scopes to classes would be too verbose)
- Output CSS, matching each scope to a single color.

Note to myself - on the conversion:

If a theme colorizes `string` and `string.quoted` differently, I should generate classes like `string` and `string-quoted`, but never `string quoted`. The CSS should use selectors like `.string-quoted`, not `.string.quoted`. Textmate themes have different order and specificity than CSS, so I'd want to avoid getting tangled in the conversion.

API wise, something like this:

```ts
interface SemanticHTMLCSSRenderer {
  generateSemanticHTML(code: string, lang: Lang): string
  generateCSSFromTheme(theme: Theme): string
}
```

**Comment by orta** [comment]
My gut says the simplest API is we allow theme to be either a `string` or `string[]`, then give each codeblock a css class with the theme name in it. Then people can use the CSS @hipstersmoothie  mentioned.

**Comment by orta** [comment]
I also joined the club of "render many times" with remark-shiki-typescript https://github.com/microsoft/TypeScript-Website/pull/1831 

 It can render multiple copies of the code, and then it's on the user to write the CSS which removes the specific theme

> ### Light / Dark Modes
> 
> If you pass more than one theme into `themes` then a codeblock will render _for each theme_ into > 
your HTML. This means that you can use CSS `display: none` on the one which shouldn't be seen.
> 
> ```ts
> const jsx = await mdx(content, {
>   filepath: "file/path/file.mdx",
>   remarkPlugins: [[remarkShikiTwoslash, { themes: ["dark-plus", "light-plus"] }]],
> })
> ```
> 
> ```css
> @media (prefers-color-scheme: light) {
>   .shiki.dark-plus {
>     display: none;
>   }
> }
> 
> @media (prefers-color-scheme: dark) {
>   .shiki.light-plus {
>     display: none;
>   }
> }
> ```

**Comment by antfu** [comment]
For the record, I made `markdown-it-shiki` using the similar approach of rendering twice:

https://github.com/antfu/markdown-it-shiki#dark-mode

**Comment by Enter-tainer** [comment]
#212 is great! Is is possible to get required CSS from existing theme?

**Comment by naiyerasif** [comment]
@Enter-tainer Take a look at docs here: https://github.com/shikijs/shiki/blob/main/docs/themes.md#theming-with-css-variables

**Comment by octref** [comment]
Hey @FredKSchott, great work! While I'm late to review the PR, since we are still pre 1.0 I'd like to get the API right. IMO hiding this functionality behind a special theme is not as good as having an explicit API. 

I think what you have is a good start. What I also want to see are:

- A standalone API in addition to `codeToHtml`. Adding option to `codeToHtml` is not good since `theme` is uesless in that case. Not sure what's a good name. Maybe `codeToParameterizedHtml` or `codeToHtmlWithCssVariables`?
- A way for users to get tokens with groups, similar to `codeToThemedTokens` but outputting token group instead of color

**Comment by orta** [comment]
I'm a bit wary about having this as a separate API codepath, interested to see how it turns out. Having a different API means that higher level abstractions like remark-shiki would need to be aware of this and expose new config vars for what is essentially a custom theme output

**Comment by octref** [comment]
@orta That's a valid concern. However I'm thinking about additional use cases like this. So far what we have are:

```
code + theme + grammar = themedTokens
themedTokens + renderer = html/svg/etc...
```

This assumes user want the tokens already colorized, and the only way for them to tweak the coloring pipeline is to write a theme. What we could also have is:

```
code + grammar = groupedTokens
```

Essentially what highlight.js does. People who want to write their own colorizer can use this API. It's also not that much – just mapping `functions/keywords` etc to a color and you are done.

I also feel `codeToHtml` should output directly embeddable HTML, not "given an option and then you need to fill in some CSS" kind of HTML.

**Comment by OskarGroth** [comment]
I'm trying out the `css-variables` theme and I see several issues with this approach:

No clear API, functionality is triggered by a theme name (`'css-variables'`)
Limited to only 12 colors
No control over color names. I'm forced to use colors like `shiki-token-punctuation` for something semantically different to make use of all 12 colors.

Instead of the hacky and limited `css-variables` theme, why not just enable the user to supply their own color remap here: https://github.com/shikijs/shiki/blob/4478d642a005b4da714743b1cd6aa5b71daf7706/packages/shiki/src/highlighter.ts#L67 and instead of the name check here https://github.com/shikijs/shiki/blob/4478d642a005b4da714743b1cd6aa5b71daf7706/packages/shiki/src/highlighter.ts#L99, add some new option on `ShikiTheme` like `cssVariables: boolean`?

Seems to me that by doing this it would unlock full theming via CSS Variables and resolve all dark-mode issues, including my problems listed above. Users would be able to swap out any colors they'd like for a CSS variable instead.

**Comment by OskarGroth** [comment]
Also the `COLOR_REPLACEMENT` map even contains errors, there is no #3 entry...

https://github.com/shikijs/shiki/blob/4478d642a005b4da714743b1cd6aa5b71daf7706/packages/shiki/src/highlighter.ts#L69

**Comment by ng-hai** [comment]
As an innocent user, I thought I could do this on `my-theme.json`

```json
{
  "tokenColors": [
    {
      "scope": ["comment"],
      "settings": {
        "foreground": "var(--color-muted)",
        "fontStyle": "italic"
      }
    },
    {
      "scope": ["constant"],
      "settings": {
        "foreground": "var(--color-pine)"
      }
    },
    {
      "scope": [
        "constant.numeric",
        "constant.language",
        "constant.charcter.escape"
      ],
      "settings": {
        "foreground": "var(--color-rose)"
      }
    },
  ]
}
```


So when Shiki processing, it will inject that value to token style `<span style="color: var(--color-muted)">variable</span>` 😓

**Comment by ng-hai** [comment]
I think I've done it, not only dark mode but also multiple themes. Inspiration by `rehype-pretty-code` and Fatih Kalifa’s [blog post](https://fatihkalifa.com/typescript-twoslash).


https://user-images.githubusercontent.com/19520278/168318199-00902981-8e8d-437f-b980-0bcf4fd42cc9.mov

**Comment by antfu** [comment]
I think maybe I found the most balanced way of doing so. You can learn more about the details at https://github.com/antfu/shikiji/pull/5

Basically, it generated inline CSS variables like:

```html
<span style="color:#1976D2;--shiki-dark:#D8DEE9">console</span>
```

That can be overridden with a short CSS snippet:

```css
@media (prefers-color-scheme: dark) {
  .shiki span {
    color: var(--shiki-dark) !important;
  }
}
```

This way we will have a perfectly working light mode output, while being able to switch to dark mode conditionally, without duplicating the content.

Would be happy to port it back to Shiki if you think this approach makes sense.

**Comment by orta** [comment]
Yeah, I think this answer is the right one - and should probably be canonically the answer 

Something I'm not certain about (from a high level) is what might happen if the two themes have different support for coloring source attributes, e.g.

```
I < 3
```
might generate something like  

```
  <code>
    <span class="line">
      <span style="color:#1976D2;--shiki-dark:#D8DEE9">I</span>
      <span style="color:#6F42C1;--shiki-dark:#ECEFF4"><</span>
      <span style="color:#6F42C1;--shiki-dark:#88C0D0">3</span>
    </span>
 </code>
```

Is it possible that there could be inconsistencies in the theme in terms of the  tokens <=> colors they support, making something like:


```
  <code>
    <span class="line">
      <span style="color:#1976D2;--shiki-dark:#D8DEE9">I</span>
      <span style="color:#6F42C1;"><</span>
      <span style="color:#6F42C1;--shiki-dark:#88C0D0">3</span>
    </span>
 </code>
```

possibly happen?

**Comment by Gerrit0** [comment]
Yes, it's absolutely possible for that to happen. TypeDoc's implementation of this uses a single class for each color, and overrides what the class does, so I've seen this quite a lot when debugging, even when highlighting with "sister" themes like Light Plus/Dark Plus

**Comment by antfu** [comment]
@orta I think `shikiji` hands it correctly. For tokens without colors, `inherit` will be rendered:

```html
<code>
  <span class="line">
    <span style="color:#1976D2;--shiki-dark:#D8DEE9">I</span>
    <span style="color:#6F42C1;--shiki-dark:inherit"><</span>
    <span style="color:#6F42C1;--shiki-dark:#88C0D0">3</span>
  </span>
</code>
```

Also it would be cases some theme output larger token, like:

```html
<code>
  <span class="line">
    <span style="color:#1976D2;">I < 3</span>
  </span>
</code>
```

In that case, `shikiji` will break those tokens into the common set of two themes.

So far I think it covers most of the edge cases.

**Comment by orta** [comment]
👍🏻

**Comment by Tachi107** [comment]
Since nobody mentioned this already: using inline html `style` attributes is somewhat undesirable, since it requires setting `style-src: 'unsafe-inline'` Content Security Policy source.

As far as I understand, the two approaches proposed in this thread rely on HTML classes and CSS variables, respectively. Since they both require the use of a separate CSS stylesheet, I'd argue that using HTML classes is better since it doesn't have the aforementioned CSP issue.

Am I missing something? Please let me know!

Thanks for your awesome work on Shiki :D
